tools/fm-editor/FeatureModelEditor.cpp

```diff
@@ -0,0 +1,312 @@
+#include "FeatureModelEditor.h"
+#include "FeatureAddDialog.h"
+#include "graph/FeatureModelGraph.h"
+#include "graph/FeatureNode.h"
+#include "ui_FeatureModelEditor.h"
+#include "vara/Feature/FeatureModel.h"
+#include "vara/Feature/FeatureModelTransaction.h"
+#include "vara/Feature/FeatureModelWriter.h"
+
+#include <QDir>
+#include <QFileDialog>
+#include <QTextStream>
+
+using Transaction = vara::feature::FeatureModelTransaction<
+    vara::feature::detail::ModifyTransactionMode>;
+using vara::feature::Feature;
+
+FeatureModelEditor::FeatureModelEditor(QWidget *Parent)
+    : QMainWindow(Parent), Ui(new Ui::FeatureModelEditor) {
+
+  Ui->setupUi(this);
+  Ui->textEdit->setReadOnly(true);
+  Highlighter = std::make_unique<QSourceHighlite::QSourceHighliter>(
+      Ui->textEdit->document());
+  Highlighter->setCurrentLanguage(QSourceHighlite::QSourceHighliter::CodeCpp);
+  QObject::connect(Ui->loadModel, &QPushButton::pressed, this,
+                   &FeatureModelEditor::loadGraph);
+  QObject::connect(Ui->actionAddFeature, &QAction::triggered, this,
+                   &FeatureModelEditor::featureAddDialog);
+  connect(Ui->actionSave, &QAction::triggered, this, &FeatureModelEditor::save);
+  connect(Ui->addSource, &QPushButton::pressed, this,
+          &FeatureModelEditor::addSource);
+  connect(Ui->addSourceFile, &QPushButton::pressed, this,
+          &FeatureModelEditor::addSourceFile);
+}
+
+/// Display the information of a Feature
+void FeatureModelEditor::loadFeature(const vara::feature::Feature *Feature) {
+  Ui->featureInfo->setText(QString::fromStdString(Feature->toString()));
+}
+
+/// Get a Feature from an Index of the TreeView and display its information.
+void FeatureModelEditor::loadFeatureFromIndex(const QModelIndex &Index) {
```

Comment:
```suggestion
void FeatureModelEditor::loadFeatureFromSelection(const QItemSelection & selection) {
  if (selection.size() != 1){
    return;
  }
  auto Index = selection.indexes().at(0);
```

To handle changes via arrow keys in TreeView